src/Http/Http/src/Builder/ApplicationBuilder.cs

```diff
@@ -129,7 +130,14 @@ public RequestDelegate Build()
                 throw new InvalidOperationException(message);
             }
 
-            context.Response.StatusCode = StatusCodes.Status404NotFound;
+            const int statusCode = StatusCodes.Status404NotFound;
+            context.Response.StatusCode = statusCode;
+
+            if (context.RequestServices.GetService<IProblemDetailsService>() is { } problemDetailsService)
+            {
+                return problemDetailsService.WriteAsync(context, isRouting: true);
+            }
```

Comment:
This should not be baked into the builder.

Reply:
Good to know. Let me see how can I implement the same behavior without it.

Reply:
We don't want problem details at this level. Something else should turn this into a problem details, like the status pages middleware.

Reply:
Yeah! I am trying to change all the routing issues to somewhere else (probably an existing Middleware) and probably we will be able to remove the `isRouting` parameter.. I will update the PR soon.

Thanks for the feedback.